proto/pbsubscribe/subscribe.proto

```diff
@@ -0,0 +1,126 @@
+    ServiceHealth = 1;
+    // ServiceHealthConnect topic contains events for any changes to service
+    // health for connect-enabled services.
+    ServiceHealthConnect = 2;
+}
+
+// SubscribeRequest used to subscribe to a topic.
+message SubscribeRequest {
+    // Topic identifies the set of events the subscriber is interested in.
+    Topic Topic = 1;
+
+    // Key is a topic-specific identifier that restricts the scope of the
+    // subscription to only events pertaining to that identifier. For example,
+    // to receive events for a single service, the service's name is
+    // specified as the key. An empty key indicates that all events in the topic
+    // are of interest.
+    string Key = 2;
+
+    // Token is the ACL token to authenticate the request. The token must have
+    // sufficient privileges to read the requested information otherwise events
+    // will be filtered, possibly resulting in an empty snapshot and no further
+    // updates sent.
+    string Token = 3;
+
+    // Index is the raft index the subscriber has already observed up to. This
+    // is zero on an initial streaming call, but then can be provided by a
+    // client on subsequent re-connections such that the full snapshot doesn't
+    // need to be resent if the client is up to date.
+    uint64 Index = 4;
+
+    // Datacenter specifies the Consul datacenter the request is targeted at.
+    // If it's not the local DC the server will forward the request to
+    // the remote DC and proxy the results back  to the subscriber. An empty
+    // string defaults to the local datacenter.
+    string Datacenter = 5;
+}
+
+// Event describes a streaming update on a subscription. Events are used both to
+// describe the current "snapshot" of the result as well as ongoing mutations to
+// that snapshot.
+message Event {
+    // Topic the event was published to
+    Topic Topic = 1;
+
+    // Key is the logical identifier for the entity that was mutated.
+    string Key = 2;
+
+    // Index is the raft index at which the mutation took place. At the top
+    // level of a subscription there will always be at most one Event per index.
+    // If multiple events are published to the same topic in a single raft
+    // transaction then the batch of events will be encoded inside a single
+    // top-level event to ensure they are delivered atomically to clients.
+    uint64 Index = 3;
+
+    // Payload is the actual event content.
+    oneof Payload {
+        // EndOfSnapshot indicates the event stream for the initial snapshot has
+        // ended. Subsequent Events delivered will be mutations to that result.
+        bool EndOfSnapshot = 5;
+
+        // EndOfEmptySnapshot indicates that the client is still up-to-date.
+        // The snapshot has ended, and was empty. The rest of the stream will be
+        // individual update events. It distinguishes between "up to date, no snapshot"
+        // and "snapshot contains zero events".
```

Comment:
Ha ha even though I wrote this originally it took me a few reads to convince myself this comment was right. I _think_ it maybe makes the point a tiny bit clearer (i.e. answers _why_ does it matter more) with a tweak like:
```suggestion
        // and "snapshot contains zero events but you should reset any old state to be blank".
```